log: better error reporting

tbg opened the pull request:
The previous code tended to lose the type when it mattered most, for example in
`errors.Wrap(unknownSentinelErr, "something")`, the output would have only an
empty string for the unknown sentinel error. This will now print its type
instead (which could still be useless, but it's better than nothing).

tbg committed:
log: better error reporting

The previous code tended to lose the type when it mattered most, for example in
`errors.Wrap(unknownSentinelErr, "something")`, the output would have only an
empty string for the unknown sentinel error. This will now print its type
instead (which could still be useless, but it's better than nothing).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19780)
<!-- Reviewable:end -->

tschottdorf commented:
Added a second commit that I think is very useful because it reports all the involved file:line tuples for any `pkg/errors`-wrapped error chains.

tbg committed:
log: extract file:line from most opaque errors in reports

This makes previously completely opaque errors such as `errors.Errorf("%s",
"foo")` much more useful by redacting them to their file:line. In particular,
most `errors.Wrap` chains should essentially emit a piecemeal stack trace.

Note that this only works for the `pkg/errors` package, but we use that in most
places (as a drop-in for the `errors` package).

dt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19780#-:-KyHq7B9-UE9JkFnqf5b:bg7igkv)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
TFTR!
